arch/power/slide_hash_power8.c

```diff
@@ -0,0 +1,71 @@
+ /* Copyright (C) 2019 Matheus Castanho <msc@linux.ibm.com>, IBM
+  * For conditions of distribution and use, see copyright notice in zlib.h
+  */
+
+#include <altivec.h>
+#include "../../zbuild.h"
+#include "../../deflate.h"
+
+static inline void slide_hash_power8_loop (deflate_state *s,
+      unsigned n_elems, Pos *table_end) __attribute__((always_inline));
+
+static void slide_hash_power8_loop(
+    deflate_state *s,
+    unsigned n_elems,
+    Pos *table_end)
+{
+    vector unsigned short vw, vm, vc, vs, *vp;
+    const vector unsigned short v0 = {0};
+    unsigned chunks;
+
+    /* Each vector register (chunk) corresponds to 128 bits == 8 Posf,
+     * so instead of processing each of the n_elems in the hash table
+     * individually, we can do it in chunks of 8 with vector instructions.
+     *
+     * This function is only called from slide_hash_power8(), and both calls
+     * pass n_elems as a power of 2 higher than 2^7, as defined by
+     * deflateInit2_(), so n_elems will always be a multiple of 8. */
+    chunks = n_elems >> 3;
+    Assert(n_elems % 8 == 0, "Weird hash table size!");
+
+    /* This type casting is safe since s->w_size is always <= 64KB
+     * as defined by deflateInit2_() and Posf == unsigned short */
+    vw[0] = (Pos) s->w_size;
+    vw = vec_splat(vw,0);
+
+    vp = (vector unsigned short *) table_end;
+
+    do {
+        /* Processing 8 elements at a time */
+        vp--;
+        vm = *vp;
+
+        /* m < s->w_size ? */
+        vc = (vector unsigned short) vec_cmplt(vm,vw);
+
+        /* calculate (m - s->w_size)
+         * Overflowed results will be discarded later  */
+        vs = vec_sub(vm,vw);
```

Comment:
Is this faster than using vec_subs()? As far as I know vec_subs() avoids the underflow.

Reply:
@mtl1979 thanks for the hint. It's been a while since I wrote this code, and I had forgotten about `vec_subs`, that will actually greatly simplify the code. I'm just testing and will push the changes shortly.

Reply:
New code already available. But I still have to update the benchmarks because the code is likely faster.

Reply:
Benchmarks updated ([link](https://github.com/zlib-ng/zlib-ng/pull/603#issuecomment-635604454)).